File: src/test/java/no/nav/syfo/services/GeografiskTilgangServiceTest.java
package no.nav.syfo.services;

import no.nav.syfo.axsys.AxsysConsumer;
import no.nav.syfo.axsys.AxsysEnhet;
import no.nav.syfo.domain.AdRoller;
import no.nav.syfo.norg2.NorgConsumer;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static no.nav.syfo.domain.AdRoller.REGIONAL;
import static no.nav.syfo.testhelper.UserConstants.NAV_ENHET_NAVN;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class GeografiskTilgangServiceTest {

    private static final String VEILEDER_UID = "Z999999";
    private static final String BRUKERS_ENHET = "brukersEnhet";
    private static final String VEILEDERS_ENHET = "veiledersEnhet";
    private static final String OVERORDNET_ENHET = "fylkeskontor";
    private static final String GEOGRAFISK_TILKNYTNING = "brukersPostnummer";

    @Mock
    private AxsysConsumer axsysConsumer;
    @Mock
    private LdapService ldapService;
    @Mock
    private NorgConsumer norgConsumer;
    @InjectMocks
    private GeografiskTilgangService geografiskTilgangService;

    @Before
    public void setup() {
        when(ldapService.harTilgang(anyString(), any())).thenReturn(false);
        when(norgConsumer.getNAVKontorForGT(GEOGRAFISK_TILKNYTNING)).thenReturn(
                BRUKERS_ENHET
        );
    }

    @Test
    public void nasjonalTilgangGirTilgang() {
        when(ldapService.harTilgang(VEILEDER_UID, AdRoller.NASJONAL.rolle)).thenReturn(true);
        assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, GEOGRAFISK_TILKNYTNING)).isTrue();
    }

    @Test
    public void utvidetTilNasjonalTilgangGirTilgang() {
        when(ldapService.harTilgang(VEILEDER_UID, AdRoller.UTVIDBAR_TIL_NASJONAL.rolle)).thenReturn(true);
        assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, GEOGRAFISK_TILKNYTNING)).isTrue();
    }

    @Test
    public void harTilgangHvisVeilederHarTilgangTilSammeEnhetSomBruker() {
        when(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
                asList(
                        new AxsysEnhet(
                                BRUKERS_ENHET,
                                NAV_ENHET_NAVN
                        ),
                        new AxsysEnhet(
                                "enHeltAnnenEnhet",
                                NAV_ENHET_NAVN
                        ))
        );
        assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, GEOGRAFISK_TILKNYTNING)).isTrue();
    }

    @Test
    public void harIkkeTilgangHvisVeilederIkkeHarTilgangTilSammeEnhetSomBruker() {
        when(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
                singletonList(new AxsysEnhet(
                        "enHeltAnnenEnhet",
                        NAV_ENHET_NAVN
                ))
        );
        assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, GEOGRAFISK_TILKNYTNING)).isFalse();
    }

    @Test
    public void harTilgangHvisRegionalTilgangOgTilgangTilEnhetensFylkeskontor() {
        when(ldapService.harTilgang(VEILEDER_UID, REGIONAL.rolle)).thenReturn(true);
        when(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
                singletonList(new AxsysEnhet(
                        VEILEDERS_ENHET,
                        NAV_ENHET_NAVN
                ))
        );
        when(norgConsumer.getOverordnetEnhetListForNAVKontor(VEILEDERS_ENHET)).thenReturn(singletonList(OVERORDNET_ENHET));
        when(norgConsumer.getOverordnetEnhetListForNAVKontor(BRUKERS_ENHET)).thenReturn(singletonList(OVERORDNET_ENHET));
        assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, GEOGRAFISK_TILKNYTNING)).isTrue();
    }

    @Test
    public void harIkkeTilgangHvisTilgangTilEnhetensFylkeskontorMenIkkeRegionalTilgang() {
        when(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
                singletonList(new AxsysEnhet(
                        OVERORDNET_ENHET,
                        NAV_ENHET_NAVN
                ))
        );
        assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, GEOGRAFISK_TILKNYTNING)).isFalse();
    }
}

File: src/test/kotlin/no/nav/syfo/geografisktilknyntning/GeografiskTilgangServiceTest.kt
package no.nav.syfo.geografisktilknyntning

import no.nav.syfo.LocalApplication
import no.nav.syfo.consumer.axsys.AxsysConsumer
import no.nav.syfo.consumer.axsys.AxsysEnhet
import no.nav.syfo.consumer.behandlendeenhet.BehandlendeEnhetConsumer
import no.nav.syfo.consumer.graphapi.GraphApiConsumer
import no.nav.syfo.consumer.norg2.NorgConsumer
import no.nav.syfo.consumer.pdl.*
import no.nav.syfo.domain.AdRoller
import no.nav.syfo.geografisktilknytning.GeografiskTilgangService
import no.nav.syfo.testhelper.UserConstants.NAV_ENHET_NAVN
import no.nav.syfo.testhelper.UserConstants.PERSON_FNR
import no.nav.syfo.testhelper.generateBehandlendeEnhet
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.context.junit.jupiter.SpringExtension

@ExtendWith(SpringExtension::class)
@SpringBootTest(classes = [LocalApplication::class])
class GeografiskTilgangServiceTest {

    @MockBean
    private lateinit var axsysConsumer: AxsysConsumer

    @MockBean
    private lateinit var behandlendeEnhetConsumer: BehandlendeEnhetConsumer

    @MockBean
    private lateinit var graphApiConsumer: GraphApiConsumer

    @MockBean
    private lateinit var norgConsumer: NorgConsumer

    @MockBean
    private lateinit var pdlConsumer: PdlConsumer

    @Autowired
    private lateinit var adRoller: AdRoller

    @Autowired
    private lateinit var geografiskTilgangService: GeografiskTilgangService

    @BeforeEach
    fun setup() {
        Mockito.`when`(norgConsumer.getNAVKontorForGT(GEOGRAFISK_TILKNYTNING)).thenReturn(
            BRUKERS_ENHET
        )
    }

    @Test
    fun nasjonalTilgangGirTilgang() {
        Mockito.`when`(graphApiConsumer.hasAccess(VEILEDER_UID, adRoller.NASJONAL)).thenReturn(true)
        Assertions.assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, PERSON_FNR)).isTrue
    }

    @Test
    fun utvidetTilNasjonalTilgangGirTilgang() {
        Mockito.`when`(graphApiConsumer.hasAccess(VEILEDER_UID, adRoller.UTVIDBAR_TIL_NASJONAL)).thenReturn(true)
        Assertions.assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, PERSON_FNR)).isTrue
    }

    @Test
    fun harTilgangHvisVeilederHarTilgangTilSammeEnhetSomBruker() {
        Mockito.`when`(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
            listOf(
                AxsysEnhet(
                    BRUKERS_ENHET,
                    NAV_ENHET_NAVN
                ),
                AxsysEnhet(
                    "enHeltAnnenEnhet",
                    NAV_ENHET_NAVN
                ))
        )
        Mockito.`when`(pdlConsumer.geografiskTilknytning(PERSON_FNR)).thenReturn(GEOGRAFISK_TILKNYTNING)

        Assertions.assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, PERSON_FNR)).isTrue
    }

    @Test
    fun harTilgangHvisVeilederHarTilgangTilSammeEnhetSomBrukerUtland() {
        val behandlendeEnhet = generateBehandlendeEnhet(BRUKERS_ENHET)
        Mockito.`when`(behandlendeEnhetConsumer.getBehandlendeEnhet(PERSON_FNR, null)).thenReturn(behandlendeEnhet)
        Mockito.`when`(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
            listOf(
                AxsysEnhet(
                    BRUKERS_ENHET,
                    NAV_ENHET_NAVN
                ),
                AxsysEnhet(
                    "enHeltAnnenEnhet",
                    NAV_ENHET_NAVN
                ))
        )
        Mockito.`when`(pdlConsumer.geografiskTilknytning(PERSON_FNR)).thenReturn(GEOGRAFISK_TILKNYTNING_UTLAND)

        Assertions.assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, PERSON_FNR)).isTrue
    }

    @Test
    fun harIkkeTilgangHvisVeilederIkkeHarTilgangTilSammeEnhetSomBrukerUtland() {
        val behandlendeEnhet = generateBehandlendeEnhet(BRUKERS_ENHET)
        Mockito.`when`(behandlendeEnhetConsumer.getBehandlendeEnhet(PERSON_FNR, null)).thenReturn(behandlendeEnhet)
        Mockito.`when`(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
            listOf(AxsysEnhet(
                "enHeltAnnenEnhet",
                NAV_ENHET_NAVN
            ))
        )
        Mockito.`when`(pdlConsumer.geografiskTilknytning(PERSON_FNR)).thenReturn(GEOGRAFISK_TILKNYTNING_UTLAND)

        Assertions.assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, PERSON_FNR)).isFalse
    }

    @Test
    fun harIkkeTilgangHvisVeilederIkkeHarTilgangTilSammeEnhetSomBruker() {
        Mockito.`when`(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
            listOf(AxsysEnhet(
                "enHeltAnnenEnhet",
                NAV_ENHET_NAVN
            ))
        )
        Mockito.`when`(pdlConsumer.geografiskTilknytning(PERSON_FNR)).thenReturn(GEOGRAFISK_TILKNYTNING)

        Assertions.assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, PERSON_FNR)).isFalse
    }

    @Test
    fun harTilgangHvisRegionalTilgangOgTilgangTilEnhetensFylkeskontor() {
        Mockito.`when`(graphApiConsumer.hasAccess(VEILEDER_UID, adRoller.REGIONAL)).thenReturn(true)
        Mockito.`when`(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
            listOf(AxsysEnhet(
                VEILEDERS_ENHET,
                NAV_ENHET_NAVN
            ))
        )
        Mockito.`when`(norgConsumer.getOverordnetEnhetListForNAVKontor(VEILEDERS_ENHET)).thenReturn(listOf(OVERORDNET_ENHET))
        Mockito.`when`(norgConsumer.getOverordnetEnhetListForNAVKontor(BRUKERS_ENHET)).thenReturn(listOf(OVERORDNET_ENHET))
        Mockito.`when`(pdlConsumer.geografiskTilknytning(PERSON_FNR)).thenReturn(GEOGRAFISK_TILKNYTNING)

        Assertions.assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, PERSON_FNR)).isTrue
    }

    @Test
    fun harIkkeTilgangHvisTilgangTilEnhetensFylkeskontorMenIkkeRegionalTilgang() {
        Mockito.`when`(axsysConsumer.enheter(VEILEDER_UID)).thenReturn(
            listOf(AxsysEnhet(
                OVERORDNET_ENHET,
                NAV_ENHET_NAVN
            ))
        )
        Mockito.`when`(pdlConsumer.geografiskTilknytning(PERSON_FNR)).thenReturn(GEOGRAFISK_TILKNYTNING)

        Assertions.assertThat(geografiskTilgangService.harGeografiskTilgang(VEILEDER_UID, PERSON_FNR)).isFalse
    }

    companion object {
        private const val VEILEDER_UID = "Z999999"
        private const val BRUKERS_ENHET = "brukersEnhet"
        private const val VEILEDERS_ENHET = "veiledersEnhet"
        private const val OVERORDNET_ENHET = "fylkeskontor"
        private val GEOGRAFISK_TILKNYTNING =
            GeografiskTilknytning(
                type = GeografiskTilknytningType.KOMMUNE,
                value = "brukersPostnummer",
            )

        private val GEOGRAFISK_TILKNYTNING_UTLAND = GeografiskTilknytning(
            type = GeografiskTilknytningType.UTLAND,
            value = "SWE",
        )
    }
}
